### PR TITLE
WorkbookEditorsHandler: Fix NPE in case name is null

### LINKED ISSUE #1275: Quick switch editor (Cmd+E) throws NPE
I20231102-1800

macOS 14.0

In a workspace with many open editors, press Cmd+E or use the command "Quick switch editor" via Cmd+3. We get NPE (check Error Log view) and the command is not executed.

```
org.eclipse.e4.ui.workbench
Error
Mon Nov 06 12:29:33 IST 2023
Execution exception for: ParameterizedCommand(Command(org.eclipse.ui.window.openEditorDropDown,Quick Switch Editor,
		Open the editor drop down list,
		Category(org.eclipse.ui.category.window,Window,null,true),
		WorkbenchHandlerServiceHandler("org.eclipse.ui.window.openEditorDropDown"),
		,,true),null) in 
	context chain: WorkbenchContext -> TrimmedWindowImpl (IDEWindow) Context -> PerspectiveImpl (org.eclipse.jdt.ui.JavaPerspective) Context -> PartImpl (org.eclipse.e4.ui.compatibility.editor)  removeOnHide org.eclipse.jdt.ui.CompilationUnitEditor (TestBug566949_1.java) Context

org.eclipse.core.commands.ExecutionException: Error executing 'org.eclipse.ui.window.openEditorDropDown': java.lang.NullPointerException: element cannot be mapped to a null key
	at org.eclipse.e4.core.commands.internal.HandlerServiceHandler.execute(HandlerServiceHandler.java:170)
	at org.eclipse.core.commands.Command.executeWithChecks(Command.java:488)
	at org.eclipse.core.commands.ParameterizedCommand.executeWithChecks(ParameterizedCommand.java:485)
	at org.eclipse.e4.core.commands.internal.HandlerServiceImpl.executeHandler(HandlerServiceImpl.java:213)
	at org.eclipse.e4.ui.bindings.keys.KeyBindingDispatcher.executeCommand(KeyBindingDispatcher.java:308)
	at org.eclipse.e4.ui.bindings.keys.KeyBindingDispatcher.press(KeyBindingDispatcher.java:580)
	at org.eclipse.e4.ui.bindings.keys.KeyBindingDispatcher.processKeyEvent(KeyBindingDispatcher.java:655)
	at org.eclipse.e4.ui.bindings.keys.KeyBindingDispatcher.filterKeySequenceBindings(KeyBindingDispatcher.java:439)
	at org.eclipse.e4.ui.bindings.keys.KeyBindingDispatcher$KeyDownFilter.handleEvent(KeyBindingDispatcher.java:96)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
	at org.eclipse.swt.widgets.Display.filterEvent(Display.java:1217)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4641)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1524)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1547)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1532)
	at org.eclipse.swt.widgets.Widget.sendKeyEvent(Widget.java:1561)
	at org.eclipse.swt.widgets.Widget.sendKeyEvent(Widget.java:1557)
	at org.eclipse.swt.widgets.Canvas.sendKeyEvent(Canvas.java:522)
	at org.eclipse.swt.widgets.Control.doCommandBySelector(Control.java:1101)
	at org.eclipse.swt.widgets.Display.windowProc(Display.java:6493)
	at org.eclipse.swt.internal.cocoa.OS.objc_msgSend(Native Method)
	at org.eclipse.swt.internal.cocoa.NSResponder.interpretKeyEvents(NSResponder.java:59)
	at org.eclipse.swt.widgets.Composite.keyDown(Composite.java:607)
	at org.eclipse.swt.widgets.Display.windowProc(Display.java:6325)
	at org.eclipse.swt.internal.cocoa.OS.objc_msgSendSuper(Native Method)
	at org.eclipse.swt.widgets.Widget.callSuper(Widget.java:236)
	at org.eclipse.swt.widgets.Widget.windowSendEvent(Widget.java:2264)
	at org.eclipse.swt.widgets.Shell.windowSendEvent(Shell.java:2511)
	at org.eclipse.swt.widgets.Display.windowProc(Display.java:6445)
	at org.eclipse.swt.internal.cocoa.OS.objc_msgSendSuper(Native Method)
	at org.eclipse.swt.widgets.Display.applicationSendEvent(Display.java:5692)
	at org.eclipse.swt.widgets.Display.applicationProc(Display.java:5832)
	at org.eclipse.swt.internal.cocoa.OS.objc_msgSend(Native Method)
	at org.eclipse.swt.internal.cocoa.NSApplication.sendEvent(NSApplication.java:117)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3986)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1155)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:342)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1046)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:155)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:645)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:342)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:552)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:171)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:152)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:203)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:136)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:402)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:255)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	at java.base/java.lang.reflect.Method.invoke(Method.java:577)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:651)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:588)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1459)
Caused by: org.eclipse.e4.core.di.InjectionException: java.lang.NullPointerException: element cannot be mapped to a null key
	at org.eclipse.e4.core.internal.di.MethodRequestor.execute(MethodRequestor.java:68)
	at org.eclipse.e4.core.internal.di.InjectorImpl.invokeUsingClass(InjectorImpl.java:302)
	at org.eclipse.e4.core.internal.di.InjectorImpl.invoke(InjectorImpl.java:236)
	at org.eclipse.e4.core.contexts.ContextInjectionFactory.invoke(ContextInjectionFactory.java:173)
	at org.eclipse.e4.core.commands.internal.HandlerServiceHandler.execute(HandlerServiceHandler.java:156)
	... 53 more
Caused by: java.lang.NullPointerException: element cannot be mapped to a null key
	at java.base/java.util.Objects.requireNonNull(Objects.java:233)
	at java.base/java.util.stream.Collectors.lambda$groupingBy$53(Collectors.java:1142)
	at java.base/java.util.stream.ReduceOps$3ReducingSink.accept(ReduceOps.java:169)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1625)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
	at org.eclipse.ui.internal.WorkbookEditorsHandler.generateColumnLabelTexts(WorkbookEditorsHandler.java:166)
	at org.eclipse.ui.internal.WorkbookEditorsHandler.getInput(WorkbookEditorsHandler.java:109)
	at org.eclipse.ui.internal.FilteredTableBaseHandler.openDialog(FilteredTableBaseHandler.java:217)
	at org.eclipse.ui.internal.FilteredTableBaseHandler.execute(FilteredTableBaseHandler.java:131)
	at org.eclipse.ui.internal.handlers.HandlerProxy.execute(HandlerProxy.java:283)
	at org.eclipse.ui.internal.handlers.E4HandlerProxy.execute(E4HandlerProxy.java:97)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	at java.base/java.lang.reflect.Method.invoke(Method.java:577)
	at org.eclipse.e4.core.internal.di.MethodRequestor.execute(MethodRequestor.java:58)
	... 57 more
```

### PR DESCRIPTION
**Pull request by sratz**
Fixes #1275.

(Regression was introduced in 03277d383e9dcd93a037640c5034a0f9a67fd854)

### CODE REVIEW

File: bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/internal/WorkbookEditorsHandler.java [hunk 1]
```diff
@@ -162,10 +164,10 @@ private List<EditorReference> getParts(WorkbenchPage page) {
 	 */
 	private Map<EditorReference, String> generateColumnLabelTexts(List<EditorReference> editorReferences) {
 		Map<EditorReference, String> editorReferenceLabelTexts = new HashMap<>(editorReferences.size());
-		Map<String, List<EditorReference>> collisionsMap = editorReferences.stream()
-				.collect(Collectors.groupingBy(r -> r.getName()));
+		Map<Optional<String>, List<EditorReference>> collisionsMap = editorReferences.stream()
```
Comment (laeubi) [inline bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/internal/WorkbookEditorsHandler.java]: maybe one should filter out `null` names altogether?

Comment (sratz) [reply bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/internal/WorkbookEditorsHandler.java]: Then they would disappear from Ctrl-E?

I guess keeping them is the safest approach for now and just fix the regression introduced by 03277d383e9dcd93a037640c5034a0f9a67fd854 .

Comment (laeubi) [reply bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/internal/WorkbookEditorsHandler.java]: What do you think about using `Objects.requireNonNullOrElse(r.getName(), "N/A");` if the only problem is that name is null (whatever that the should mean to the user).

Comment (sratz) [reply bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/internal/WorkbookEditorsHandler.java]: I think I have a better solution:

The actual intended text to be used in Ctrl-E (if unique), is `ref.getTitle()`, 
https://github.com/eclipse-platform/eclipse.platform.ui/blob/5b623809b68f069f0df2b6ad27f03af3b0c554f9/bundles/org.eclipse.ui.workbench/Eclipse%20UI/org/eclipse/ui/internal/FilteredTableBaseHandler.java#L821

which is already null-safe (returns `""` in that case):
https://github.com/eclipse-platform/eclipse.platform.ui/blob/5b623809b68f069f0df2b6ad27f03af3b0c554f9/bundles/org.eclipse.ui.workbench/Eclipse%20UI/org/eclipse/ui/internal/EditorReference.java#L227-L240

But in 03277d383e9dcd93a037640c5034a0f9a67fd854 I have inadvertently used `getName()` to populate the `collisionMap`.

It should simply also use `getTitle()`.

**Comment by sratz** [comment]
Any objections to merge like this? (I don't know why the build is failing, but it's definitely unrelated).

**Review by BeckerWdf — APPROVED** [review]

**Review by laeubi — APPROVED** [review]
Looks good

**Review by iloveeclipse — CHANGES_REQUESTED** [review]
We can't merge, API tools are broken. @laeubi : please take a look.

See https://ci.eclipse.org/platform/job/eclipse.platform.ui/job/PR-1286/4/console

```
09:45:32.437 [ERROR] Failed to execute goal org.eclipse.tycho:tycho-apitools-plugin:4.0.4:verify (verify) on project org.eclipse.ui.workbench: There are API errors:
09:45:32.437 [ERROR] Eclipse UI/org/eclipse/ui/AbstractSourceProvider.java:31 The type org.eclipse.ui.AbstractSourceProvider has been removed from org.eclipse.ui.workbench_3.131.0
09:45:32.438 [ERROR] Eclipse UI/org/eclipse/ui/ActiveShellExpression.java:32 The type org.eclipse.ui.ActiveShellExpression has been removed from org.eclipse.ui.workbench_3.131.0
09:45:32.438 [ERROR] Eclipse UI/org/eclipse/ui/BasicWorkingSetElementAdapter.java:58 The type org.eclipse.ui.BasicWorkingSetElementAdapter has been removed from org.eclipse.ui.workbench_3.131.0
09:45:32.438 [ERROR] Eclipse UI/org/eclipse/ui/ExtensionFactory.java:46 The type org.eclipse.ui.ExtensionFactory has been removed from org.eclipse.ui.workbench_3.131.0
09:45:32.438 [ERROR] Eclipse UI/org/eclipse/ui/IActionBars.java:66 The type org.eclipse.ui.IActionBars has been removed from org.eclipse.ui.workbench_3.131.0
09:45:32.438 [ERROR] Eclipse UI/org/eclipse/ui/IActionBars2.java:24 The type org.eclipse.ui.IActionBars2 has been removed from org.eclipse.ui.workbench_3.131.0
09:45:32.438 [ERROR] Eclipse UI/org/eclipse/ui/IActionDelegate.java:53 The type org.eclipse.ui.IActionDelegate has been removed from org.eclipse.ui.workbench_3.131.0
09:45:32.438 [ERROR] Eclipse UI/org/eclipse/ui/IActionDelegate2.java:42 The type org.eclipse.ui.IActionDelegate2 has been removed from org.eclipse.ui.workbench_3.131.0
09:45:32.438 [ERROR] Eclipse UI/org/eclipse/ui/IActionDelegateWithEvent.java:34 The type org.eclipse.ui.IActionDelegateWithEvent has been removed from org.eclipse.ui.workbench_3.131.0
09:45:32.438 [ERROR] Eclipse UI/org/eclipse/ui/IActionFilter.java:52 The type org.eclipse.ui.IActionFilter has been removed from org.eclipse.ui.workbench_3.131.0
09:45:32.438 [ERROR] Eclipse UI/org/eclipse/ui/IAggregateWorkingSet.java:30 The type org.eclipse.ui.IAggregateWorkingSet has been removed from org.eclipse.ui.workbench_3.131.0
09:45:32.438 [ERROR] Eclipse UI/org/eclipse/ui/IContainmentAdapter.java:22 The type 

...
```

**Comment by jukzi** [comment]
> We can't merge, API tools are broken. @laeubi : please take a look.

see https://github.com/eclipse-equinox/p2/pull/382#issuecomment-1801611393
currently it helps to add the default 'output.. = bin/' to build.properties.

File: bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/internal/WorkbookEditorsHandler.java [hunk 2]
```diff
@@ -163,7 +163,7 @@ private List<EditorReference> getParts(WorkbenchPage page) {
 	private Map<EditorReference, String> generateColumnLabelTexts(List<EditorReference> editorReferences) {
 		Map<EditorReference, String> editorReferenceLabelTexts = new HashMap<>(editorReferences.size());
 		Map<String, List<EditorReference>> collisionsMap = editorReferences.stream()
-				.collect(Collectors.groupingBy(r -> r.getName()));
+				.collect(Collectors.groupingBy(r -> r.getTitle()));
```
Comment (jukzi) [inline bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/internal/WorkbookEditorsHandler.java]: How about replacing null with empty string?

Comment (sratz) [reply bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/internal/WorkbookEditorsHandler.java]: `getTitle()` already does that and does not return null

Comment (jukzi) [reply bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/internal/WorkbookEditorsHandler.java]: that can return descriptor.getLabel() that can return configurationElement.getAttribute(IWorkbenchRegistryConstants.ATT_NAME); that can return null

**Review by jukzi — APPROVED** [review]

**Comment by iloveeclipse** [comment]
> see [eclipse-equinox/p2#382 (comment)](https://github.com/eclipse-equinox/p2/pull/382#issuecomment-1801611393) currently it helps to add the default 'output.. = bin/' to build.properties.

1) The problem is still there. Please fix that (even if it is not introduced by you).
2) M3 is currently in the process to be declared. **No** merges are expected before that.
3) Once M3 is officially declared, we could try to merge (assuming 1) is solved and build is green), but we will need a positive vote from a project lead to do that.

**Comment by jukzi** [comment]
> 1. The problem is still there. Please fix that (even if it is not introduced by you).

it's tychos problem https://github.com/eclipse-tycho/tycho/issues/3019

File: bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/internal/WorkbookEditorsHandler.java (hunk 2 above)
Comment (sratz) [reply bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/internal/WorkbookEditorsHandler.java]: True, I have added another Util.safeString() guard around it.

**Comment by sratz** [comment]
@iloveeclipse: Any objections to merge for RC1?

**Review by iloveeclipse — APPROVED** [review]
LGTM

**Comment by laeubi** [comment]
@vogella @merks we need a PL/PMC approval for RC1 I think its worth to be considered.
@iloveeclipse build is green now except flaky `VirtualTableViewerTest` what I think is not affected by this change.

**Comment by merks** [comment]
PMC +1

I do generally trust that folks will use their discretion to fix things that really ought to be fixed and to avoid changing things that are risk.

**Pull request merged**